src/shared/rm-rf.c

```diff
@@ -115,6 +115,57 @@ int fstatat_harder(int dfd,
         return 0;
 }
 
+static int openat_harder(int dfd, const char *path, int open_flags, RemoveFlags remove_flags, mode_t *ret_old_mode) {
+        _cleanup_close_ int pfd = -EBADF, fd = -EBADF;
+        mode_t old_mode;
+        int r;
+
+        assert(dfd >= 0 || dfd == AT_FDCWD);
+        assert(path);
+        assert(open_flags & O_DIRECTORY);
```

Comment:
why this check btw? this funciton seems generic enough to be fine for any dentry type. seems like a red herring to add here

Reply:
Currently, `patch_dirfd_mode()` works only for directory. Hence, it does not work if `path` points to a non-directory file.
In the revised version, the assertion is dropped, but if `O_DIRECTORY` is not set, then this is mostly equivalent to `openat()`.